providers/common/kdfs/tls1_prf.c

```diff
@@ -75,137 +83,158 @@ struct evp_kdf_impl_st {
+        ERR_raise(ERR_LIB_PROV, PROV_R_MISSING_SECRET);
+        return 0;
+    }
+    if (ctx->seedlen == 0) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_MISSING_SEED);
+        return 0;
+    }
+    return tls1_prf_alg(ctx->md, ctx->sec, ctx->seclen,
+                        ctx->seed, ctx->seedlen,
+                        key, keylen);
+}
 
-    case EVP_KDF_CTRL_SET_TLS_SECRET:
-        p = va_arg(args, const unsigned char *);
-        len = va_arg(args, size_t);
-        OPENSSL_clear_free(impl->sec, impl->seclen);
-        impl->sec = OPENSSL_memdup(p, len);
-        if (impl->sec == NULL)
+static int kdf_tls1_prf_set_ctx_params(void *vctx, const OSSL_PARAM params[])
+{
+    const OSSL_PARAM *p;
+    TLS1_PRF *ctx = vctx;
+    EVP_MD *md;
+    const char *properties = NULL;
+    int first_seed = 1;
+
+    /* Grab search properties, this should be before the digest lookup */
+    if ((p = OSSL_PARAM_locate_const(params, OSSL_KDF_PARAM_PROPERTIES))
+        != NULL) {
+        if (p->data_type != OSSL_PARAM_UTF8_STRING)
             return 0;
-
-        impl->seclen = len;
-        return 1;
-
-    case EVP_KDF_CTRL_RESET_TLS_SEED:
-        OPENSSL_cleanse(impl->seed, impl->seedlen);
-        impl->seedlen = 0;
-        return 1;
-
-    case EVP_KDF_CTRL_ADD_TLS_SEED:
-        p = va_arg(args, const unsigned char *);
-        len = va_arg(args, size_t);
-        if (len == 0 || p == NULL)
-            return 1;
-
-        if (len > (TLS1_PRF_MAXBUF - impl->seedlen))
+        properties = p->data;
+    }
+    /* Handle aliasing of digest parameter names */
+    if ((p = OSSL_PARAM_locate_const(params, OSSL_KDF_PARAM_DIGEST)) != NULL) {
+        if (p->data_type != OSSL_PARAM_UTF8_STRING)
             return 0;
+        md = EVP_MD_fetch(PROV_LIBRARY_CONTEXT_OF(ctx->provctx), p->data,
+                          properties);
+        if (md == NULL) {
+            ERR_raise(ERR_LIB_PROV, PROV_R_INVALID_DIGEST);
+            return 0;
+        }
+        ctx->md = md;
+    }
 
-        memcpy(impl->seed + impl->seedlen, p, len);
-        impl->seedlen += len;
-        return 1;
-
-    default:
-        return -2;
+    if ((p = OSSL_PARAM_locate_const(params, OSSL_KDF_PARAM_SECRET)) != NULL) {
+        OPENSSL_clear_free(ctx->sec, ctx->seclen);
+        ctx->sec = NULL;
+        if (!OSSL_PARAM_get_octet_string(p, (void **)ctx->sec, 0, &ctx->seclen))
+            return 0;
     }
+    /* The seed fields concatenate, so process them all */
+    for (p = params; p->key != NULL; p++)
+        if (strcmp(p->key, OSSL_KDF_PARAM_SEED) == 0) {
+            size_t sz = 0;
+            const unsigned char *q;
+
+            if (first_seed) {
+                OPENSSL_cleanse(ctx->seed, ctx->seedlen);
+                ctx->seedlen = 0;
+                first_seed = 0;
+            }
+            q = ctx->seed + ctx->seedlen;
+            if (p->data_size != 0
+                && p->data != NULL
+                && !OSSL_PARAM_get_octet_string(p, (void **)&q,
```

Comment:
This is not very readable with params.. (Not much we can do at this point)

Reply:
Hmmm...  this could be slightly rewritten:

```
    /* Several seeds may be provided, concatenate them all */
    if ((p = OSSL_PARAM_locate_const(param, OSSL_KDF_PARAM_SEED)) != NULL) {
        OPENSSL_cleanse(ctx->seed, ctx->seedlen);
        ctx->seedlen = 0;
        for (; p != NULL; p = OSSL_PARAM_locate_const(p + 1, OSSL_KDF_PARAM_SEED)) {
            const void *q = ctx->seed + ctx->seedlen;

            if (p->data_size != 0
                && p->data == NULL
                && !OSSL_PARAM_get_octet_string(p, &q,
                                                TLS1_PRF_MAXBUF - ctx->seedlen,
                                                &sz))
                return 0;
            ctx->seedlen += sz;
        }
```

Reply:
I thought about something along these lines at the time, I don't know why I did do it.

Reply:
Changed to locate each in turn here and for hkdf.

This breaks the documentation that OSSL_PARAMs are stored/handled in any order.

Reply:
Mind pointing out the text that says so?  I'm trying to look, but cannot find anything that says this categorically.

Reply:
I can't quickly find it.  I remember writing it into the documentation but it might not have survives the review process.

A third party provider is free to process params however it wishes so assuming an ordering is fragile.  I don't see a reasonable alternative though (..._SEED_1, ..._SEED_2, ..._SEED_3, ... is nasty but would work).

Reply:
ordering requires seperate calls.

Reply:
Or accepting that OSSL_PARAM arrays are processed in order when duplicate keys are present.